_posts/2025-06-18-hibernate7-on-quarkus.adoc

```diff
@@ -0,0 +1,117 @@
+
+[source,xml]
+----
+<dependency>
+       <groupId>jakarta.data</groupId>
+       <artifactId>jakarta.data-api</artifactId>
+ </dependency>
+----
+
+Here’s an example of how a simple repository can be written:
+
+[source,java]
+----
+@Repository
+public interface Library {
+
+    @Find
+    Optional<Book> byIsbn(String isbn);
+
+    @Query("""
+            select b.isbn, b.title, listagg(a.name, ' & ')
+            from Book b join b.authors a
+            group by b
+            order by b.isbn
+            """)
+    List<Summary> summarize();
+}
+----
+
+This topic deserves a deeper dive, so let us know if you're interested, as we could provide more content.
+
+In the meantime, you can always refer to the https://quarkus.io/version/main/guides/hibernate-orm#jakarta-data-2[dedicated Quarkus guide] to get started quickly, and to the https://docs.jboss.org/hibernate/orm/7.0/repositories/html_single/Hibernate_Data_Repositories.html[corresponding documentation in Hibernate ORM] for more advanced usage.
+
+
+==== New Restrictions API
+After the deprecation of the Hibernate Criteria API, developers were still missing its simplicity, so the Hibernate team introduced a new https://docs.jboss.org/hibernate/orm/7.0/introduction/html_single/Hibernate_Introduction.html#restrictions-and-ordering[Restriction API] that even has new features, such as the possibility to further restrict an already written query using JPQL/HQL.
+
+[source,java]
+----
+List<Book> books =
+    SelectionSpecification.create(Book.class,
+        """
+            from Book where discontinued = false
+        """)
+        .restrict(Restriction.startsWith(Book_.title, "hibernate"))
+        .sort(Order.desc(Book_.title))
+        .createQuery(session)
+        .setPage(Page.first(50))
+        .getResultList();
+----
+
+This feature can also be used with https://docs.jboss.org/hibernate/orm/7.0/repositories/html_single/Hibernate_Data_Repositories.html#dynamic-restrictions[Hibernate Data Repositories] (the Hibernate implementation of the Jakarta Data API), and create a repository that allows filtering without having to write any JPQL/HQL code:
+
+```java
+@Find
+List<Book> books(Restriction<Book> restriction,
+                 Order<Book> order);
+```
+
+When user will call the method, they can pass the `Restriction` objects to filter the wanted book.
+
+```java
+var books =
+       library.books(Restriction.contains(Book_.title, "Hibernate"),
+                     Order.of(_Book.title.ascIgnoreCase(),
+                              _Book.isbn.asc()));
+```
+
+=== Hibernate Reactive together with Hibernate ORM
+
+A long-awaited feature is the ability to https://github.com/quarkusio/quarkus/issues/13425[mix Hibernate ORM and Hibernate Reactive extensions] in the same Quarkus application. Without this feature, making the two extensions coexist required workarounds that are now unnecessary as, since Quarkus 3.2.4, it is now possible to mix the two.
```

Comment:
```suggestion
A long-awaited feature is the ability to https://github.com/quarkusio/quarkus/issues/13425[mix Hibernate ORM and Hibernate Reactive extensions] in the same Quarkus application. Without this feature, making the two extensions coexist required workarounds that are now unnecessary as, since Quarkus 3.2.4, it's now possible to mix the two.
```